### ISSUE TITLE
use pino-noir to redact cookies

### ISSUE THREAD

**Issue by mcollina** [pinojs/pino-http#13]
As titled, probably a good security measure

**Comment by novemberborn** [comment]
Perhaps Pino should support *combining* serializers with redactors. That is, first call `stdSerializers.res`, then apply any `res.*` redactors to the result. This would require building redactor support into Pino itself.

Redacting cookies from the response seems tricky, since they'll be deep inside the `headers` string. `pino-noir` doesn't let you redact string contents.

**Comment by davidmarkclements** [comment]
I'd like pino core to remain laser focused on pure logging - I think we can do this without putting redaction in core

Redaction is a subset of serialization, we can look at this as nested serialization - or maybe a chain or serializers. 

We own pino-noir, there's no problems extending it to suit this case

If pino-noir accepts a pre-existing serializer object (which we create in pino-http), we can combine the serializers accordingly

Great point on the headers string, we can extend pino noir to accept a replace regex for strings - however this may cause some perf slowdown - the other option is to alter pino to expose the headers object instead of the headers string (this would make it consistent with the req serializer) 

This will of course mean serializing an object, which will also add overhead

cc @mcollina

**Comment by novemberborn** [comment]
> I'd like pino core to remain laser focused on pure logging - I think we can do this without putting redaction in core

👍 

> Redaction is a subset of serialization, we can look at this as nested serialization - or maybe a chain or serializers.
> 
> We own pino-noir, there's no problems extending it to suit this case
> 
> If pino-noir accepts a pre-existing serializer object (which we create in pino-http), we can combine the serializers accordingly

Yea that sounds like the right approach.

Let me know if I can help.

**Comment by mcollina** [comment]
@novemberborn go ahead and send PRs :).

**Comment by novemberborn** [comment]
@mcollina 😄  Let's hope I find some time. I'm using `pino` in a client project but will need to see if dropping in a different library gets the job done more easily for them…

> Great point on the headers string, we can extend pino noir to accept a replace regex for strings - however this may cause some perf slowdown - the other option is to alter pino to expose the headers object instead of the headers string (this would make it consistent with the req serializer)
> 
> This will of course mean serializing an object, which will also add overhead

@mcollina what direction do you prefer here? I like the consistency of exposing the headers object, though I haven't checked if one is actually available.

**Comment by mcollina** [comment]
This one is way faster than anything else :). 

I'm missing a bit of context on the headers. Maybe I've coded something in a hurry, I think we should change pino to expose the headers object.
Or at least change this.

**Comment by davidmarkclements** [comment]
yeah you've summarised the context - we expose _headers which is the raw string, but only for the res serializer - req exposes headers object. If we expose the headers object, implementation for this gets much easier

**Comment by jsumners** [comment]
I suspect this is related https://github.com/pinojs/pino-noir/issues/6
